flocker/testtools/_flaky.py

```diff
@@ -22,10 +37,179 @@ def flaky(jira_key):
+        run_test = self._run_test_factory(case, *self._args, **self._kwargs)
+        return run_test._run_prepared_result(result)
+
+    def _run_flaky_test(self, case, result, flaky):
+        """
+        Run a test that has been decorated with the `@flaky` decorator.
+
+        :param TestCase case: A ``testtools.TestCase`` to run.
+        :param TestResult result: A ``TestResult`` object that conforms to the
+            testtools extended result interface.
+        :param _FlakyAnnotation flaky: A description of the conditions of
+            flakiness.
+
+        :return: A ``TestResult`` with the result of running the flaky test.
+        """
+        successes = 0
+        results = []
+
+        while successes < flaky.min_passes and len(results) < flaky.max_runs:
+            was_successful, details = self._attempt_test(case)
+            if was_successful:
+                successes += 1
+            results.append(details)
+
+        flaky_data = flaky.to_dict()
+        flaky_data.update({'runs': len(results), 'passes': successes})
+        flaky_details = {
+            'flaky': text_content(pformat(flaky_data)),
+        }
+
+        details = _combine_details([flaky_details] + results)
+        result.startTest(case)
+        if successes >= flaky.min_passes:
+            result.addSuccess(case, details=details)
+        else:
+            # XXX: How are we going to report on tests that sometimes fail,
+            # sometimes error, sometimes skip? Currently we just error.
+            result.addError(case, details=details)
+        result.stopTest(case)
+        return result
+
+    def _attempt_test(self, case):
+        """
+        Run 'case' with a temporary result.
+
+        :param testtools.TestCase case: The test to run.
+
+        :return: a tuple of ``(success, details)``, where ``success`` is
+            a boolean indicating whether the test was successful or not
+            and ``details`` is a dictionary of testtools details.
+        """
+        tmp_result = testtools.TestResult()
+        self._run_test(case, tmp_result)
+        details = pmap(case.getDetails())
+        _reset_case(case)
+        return tmp_result.wasSuccessful(), details
+
+
+def _combine_details(detailses):
+    """
+    Take a sequence of details dictionaries and combine them into one.
+    """
+    # XXX: Only necessary becaause testtools's `gather_details` is perversely
+    # mutatey.
+    result = {}
+    for details in detailses:
+        gather_details(details, result)
+    return pmap(result)
+
+
+def _reset_case(case):
+    """
+    Reset ``case`` so it can be run again.
+    """
+    # XXX: Alternative approach is to use clone_test_with_new_id, rather than
+    # resetting the same test case.
+    # Don't want details from last run.
+    case.getDetails().clear()
+    # https://github.com/testing-cabal/testtools/pull/165/ fixes this.
+    case._TestCase__setup_called = False
+    case._TestCase__teardown_called = False
```

Comment:
testing-cabal/testtools#165 as currently written will at least mitigate the problem. I hope that before it lands it will fix it entirely.